vcls/magento2/default.vcl

```diff
@@ -0,0 +1,229 @@
+
+# For SSL offloading, pass the following header in your proxy server or load balancer: '/* {{ ssl_offloaded_header }} */: https'
+
+backend default none;
+
+probe p1 {
+        .url = "/pub/health_check.php";
+        .timeout = 2s;
+        .interval = 5s;
+        .window = 10;
+        .threshold = 5;
+}
+
+acl purge {
+/* {{ ips }} */
+}
+
+sub vcl_init {
+    new dir = goto.dns_director("{{ host }}:{{ port }}",
+                                probe = p1,
+                                first_byte_timeout = 10m);
+}
+
+sub vcl_recv {
+    set req.backend_hint = dir.backend();
+
+    if (req.method == "PURGE") {
+        if (client.ip !~ purge) {
+            return (synth(405, "Method not allowed"));
+        }
+        # To use the X-Pool header for purging varnish during automated deployments, make sure the X-Pool header
+        # has been added to the response in your backend server config. This is used, for example, by the
+        # capistrano-magento2 gem for purging old content from varnish during it's deploy routine.
+        if (!req.http.X-Magento-Tags-Pattern && !req.http.X-Pool) {
+            return (synth(400, "X-Magento-Tags-Pattern or X-Pool header required"));
+        }
+        if (req.http.X-Magento-Tags-Pattern) {
+            set req.http.X-Magento-Tags-Pattern = regsuball(req.http.X-Magento-Tags-Pattern, "\(\(\^\|\,\)", " ");
+            set req.http.X-Magento-Tags-Pattern = regsuball(req.http.X-Magento-Tags-Pattern, "\(\,\|\$\)\)", "");
+            set req.http.X-Magento-Tags-Pattern = regsuball(req.http.X-Magento-Tags-Pattern, "\|", ",");
+            ykey.purge_keys(req.http.X-Magento-Tags-Pattern);
+        }
+        if (req.http.X-Pool) {
+            set req.http.X-Pool = regsuball(req.http.X-Pool, "\(\(\^\|\,\)", " ");
+            set req.http.X-Pool = regsuball(req.http.X-Pool, "\(\,\|\$\)\)", "");
+            set req.http.X-Pool = regsuball(req.http.X-Pool, "\|", ",");
+            ykey.namespace("pool");
+            ykey.purge_keys(req.http.X-Pool);
+            ykey.namespace_reset();
+        }
+        return (synth(200, "Purged"));
+    }
+
+    if (req.method != "GET" &&
+        req.method != "HEAD" &&
+        req.method != "PUT" &&
+        req.method != "POST" &&
+        req.method != "TRACE" &&
+        req.method != "OPTIONS" &&
+        req.method != "DELETE") {
+          /* Non-RFC2616 or CONNECT which is weird. */
+          return (pipe);
+    }
+
+    # We only deal with GET and HEAD by default
+    if (req.method != "GET" && req.method != "HEAD") {
+        return (pass);
+    }
+
+    # Bypass customer, shopping cart, checkout
+    if (req.url ~ "/customer" || req.url ~ "/checkout") {
+        return (pass);
+    }
+
+    # Bypass health check requests
+    if (req.url ~ "^/(pub/)?(health_check.php)$") {
+        return (pass);
+    }
+
+    # if the backend is healthy, limit the grace
+    if (std.healthy(req.backend_hint)) {
+        set req.grace = /* {{ grace_period }} */s;
```

Comment:
Use this as the `beresp.grace` and nuke all the dynamic grace logic.

Reply:
check the upstream product first

Reply:
Just to clarify, lowering the grace value based on `std.healthy()` introduces a chance to get a 503 since there is a +10s lag between when a backend goes down and this VCL logic picks it up. Its highly recommended to switch this to a normal grace + stale-on-error.

This VCL is a common pattern which should be avoided, `std.healthy()` is not 100% accurate.